tools/cn-tree-simulator/cn/tree.go

```diff
@@ -0,0 +1,214 @@
+/* SPDX-License-Identifier: Apache-2.0
+ * Copyright (C) 2022 Micron Technology, Inc. All rights reserved.
+ */
+
+package cn
+
+import (
+	"container/list"
+	"encoding/json"
+	"fmt"
+)
+
+type Tree struct {
+	params *TreeParams
+	stats  treeStatistics
+
+	root rootNode
+}
+
+type node struct {
+	children *list.List
+	kvsets   *list.List
+}
+
+type rootNode struct {
+	node
+
+	params *RootNodeParams
+	stats  rootNodeStatistics
+}
+
+type leafNode struct {
+	node
+
+	params *LeafNodeParams
+	stats  leafNodeStatistics
+}
+
+func NewTree(leaves uint, initialSize uint64, params *Params) *Tree {
+	tree := Tree{
+		params: &params.Tree,
+		root: rootNode{
+			node: node{
+				children: list.New(),
+				kvsets:   list.New(),
+			},
+			params: &params.RootNode,
+		},
+	}
+
+	if initialSize != 0 {
+		for i := uint(0); i < leaves; i++ {
+			size := initialSize / uint64(leaves)
+			node := &leafNode{
+				node:   node{kvsets: list.New()},
+				params: &params.LeafNode,
+			}
+
+			tree.root.children.PushBack(node)
+
+			node.kvsets.PushFront(newKvSet(size, false))
+		}
+	}
+
+	return &tree
+}
+
+func (t *Tree) Size() uint64 {
+	total := t.root.size()
+
+	for leaf := t.root.children.Front(); leaf != nil; leaf = leaf.Next() {
+		for kvset := leaf.Value.(*leafNode).kvsets.Front(); kvset != nil; kvset = kvset.Next() {
+			total += kvset.Value.(*kvSet).size
+		}
+	}
+
+	return total
+}
+
+func (t *Tree) Ingest(data uint64) {
+	t.root.kvsets.PushFront(newKvSet(data, false))
+
+	t.stats.ingests++
+	t.stats.ingested += data
+	t.stats.written += data
+
+	if t.root.stats.spills == uint(0) && t.root.size() < t.root.params.InitialSpillSize {
+		return
+	}
+
+	if t.root.size() < t.root.params.SpillSize {
+		return
+	}
+
+	t.stats.written += t.root.size()
+	t.root.spill()
```

Comment:
Seems like t.stats.written should be updated in t.root.spill().

Reply:
Seems reasonable

Reply:
Now I remember. I don't have access to `t` in that spill function. I think in another comment, you wanted to keep stats local to node, so I will try to rework.

Reply:
I feel like trees will have config info that may be worth carrying around.  I would've violated encapsulation rules and passed the tree object around - or let each node have a ref back to the tree object.